contracts/test/v0.8/dev/keeper2_0/KeeperRegistrar2_0.test.ts

```diff
@@ -549,6 +549,93 @@ describe('KeeperRegistrar2_0', () => {
     })
   })
 
+  describe('#registerUpkeep', () => {
+    it('reverts with empty message if amount sent is not available in LINK allowance', async () => {
+      await evmRevert(
+        registrar
+          .connect(someAddress)
+          .registerUpkeep(
+            upkeepName,
+            emptyBytes,
+            mock.address,
+            executeGas,
+            await admin.getAddress(),
+            emptyBytes,
+            amount,
+          ),
+        '',
+      )
+    })
+
+    it('reverts if the amount passed in data is less than configured minimum', async () => {
+      await registrar
+        .connect(registrarOwner)
+        .setRegistrationConfig(
+          autoApproveType_ENABLED_ALL,
+          maxAllowedAutoApprove,
+          registry.address,
+          minUpkeepSpend,
+        )
+
+      // amt is one order of magnitude less than minUpkeepSpend
+      const amt = BigNumber.from('100000000000000000')
+
+      await evmRevert(
+        registrar
+          .connect(someAddress)
+          .registerUpkeep(
+            upkeepName,
+            emptyBytes,
+            mock.address,
+            executeGas,
+            await admin.getAddress(),
+            emptyBytes,
+            amt,
+          ),
+        'InsufficientPayment()',
+      )
+    })
+
+    it('Auto Approve ON - registers an upkeep on KeeperRegistry instantly and emits both RegistrationRequested and RegistrationApproved events', async () => {
+      //set auto approve ON with high threshold limits
+      await registrar
+        .connect(registrarOwner)
+        .setRegistrationConfig(
+          autoApproveType_ENABLED_ALL,
+          maxAllowedAutoApprove,
+          registry.address,
+          minUpkeepSpend,
+        )
+
+      await linkToken.connect(requestSender).approve(registrar.address, amount)
+
+      const tx = await registrar
+        .connect(requestSender)
+        .registerUpkeep(
+          upkeepName,
+          emptyBytes,
+          mock.address,
+          executeGas,
+          await admin.getAddress(),
+          emptyBytes,
+          amount,
+        )
+      assert.equal((await registry.getState()).state.numUpkeeps.toNumber(), 1) // 0 -> 1
+
+      //confirm if a new upkeep has been registered and the details are the same as the one just registered
+      const [id] = await registry.getActiveUpkeepIDs(0, 1)
+      const newupkeep = await registry.getUpkeep(id)
```

Comment:
can you actually pick the upkeep ID from tx, instead of this logic?

Reply:
also another test case would be good with auto approve off and assertion that returned upkeep ID is zero

Reply:
I might need some help with that. I tried doing so and the solution didn't come to me. I really wanted to get the returned id from the `registerUpkeep` function and compare it do the one above.

How do you extract the returned value from a `ContractTransaction`?

Reply:
Discussed offline, not trivial to do